File: crates/nil-core/src/world/military/maneuver.rs
// Copyright (C) Call of Nil contributors
// SPDX-License-Identifier: AGPL-3.0-only

use crate::error::Result;
use crate::military::maneuver::ManeuverRequest;
use crate::world::World;

impl World {
  pub fn request_maneuver(&mut self, request: &ManeuverRequest) -> Result<()> {
    // TODO: Should a player be allowed to attack themselves?
    self.military.request_maneuver(request)?;
    let player_a = self.city(request.origin)?.player();
    let player_b = self.city(request.destination)?.player();

    if let Some(player_a) = &player_a {
      self.emit_military_updated(player_a.clone());
    }

    if player_b != player_a
      && let Some(player_b) = &player_b
    {
      self.emit_military_updated(player_b.clone());
    }

    Ok(())
  }
}

File: app/src/types/core/military/maneuver.ts
// Copyright (C) Call of Nil contributors
// SPDX-License-Identifier: AGPL-3.0-only

import type { Ruler } from '@/types/core/ruler';
import type { Coord } from '@/types/core/continent';
import type { Resources } from '@/types/core/resources';
import type { ArmyId, ArmyPersonnel } from '@/types/core/military/army';

export interface Maneuver {
  readonly id: ManeuverId;
  readonly origin: Coord;
  readonly destination: Coord;
  readonly army: ArmyId;
  readonly kind: ManeuverKind;
  readonly direction: ManeuverDirection;
  readonly state: ManeuverState;
  readonly speed: number;
  readonly hauledResources: Option<ManeuverHaul>;
}

export type ManeuverId = string;

export type ManeuverKind = 'attack' | 'support';

export type ManeuverDirection = 'going' | 'returning';

export type ManeuverDistance = number;

export type ManeuverState = ManeuverStateDone | ManeuverStatePending;

export interface ManeuverStateDone {
  readonly kind: 'done';
}

export interface ManeuverStatePending {
  readonly kind: 'pending';
  readonly distance: ManeuverDistance;
}

export interface ManeuverHaul {
  readonly ruler: Ruler;
  readonly resources: Resources;
}

export interface ManeuverRequest {
  readonly kind: ManeuverKind;
  readonly origin: Coord;
  readonly destination: Coord;
  readonly personnel: ArmyPersonnel;
}
